tests/components/ecoforest/test_config_flow.py

```diff
@@ -0,0 +1,120 @@
+from homeassistant.components.ecoforest.const import DOMAIN
+from homeassistant.core import HomeAssistant
+from homeassistant.data_entry_flow import FlowResultType
+
+pytestmark = pytest.mark.usefixtures("mock_setup_entry")
+
+
+async def test_form(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock, mock_device
+) -> None:
+    """Test we get the form."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
+
+    with patch(
+        "pyecoforest.api.EcoforestApi.get",
+        return_value=mock_device,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.1.1.1",
+                "username": "test-username",
+                "password": "test-password",
+            },
+        )
+        await hass.async_block_till_done()
+
+    assert result2["type"] == FlowResultType.CREATE_ENTRY
+    assert result2["title"] == "Ecoforest 1234"
+    assert result2["data"] == {
+        "host": "1.1.1.1",
+        "username": "test-username",
+        "password": "test-password",
+    }
+    assert len(mock_setup_entry.mock_calls) == 1
+
+
+async def test_form_host_already_configured(
+    hass: HomeAssistant, mock_setup_entry: AsyncMock, config_entry, mock_device
+) -> None:
+    """Test host already exists."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+    assert result["type"] == FlowResultType.FORM
+    assert result["errors"] == {}
+
+    with patch(
+        "pyecoforest.api.EcoforestApi.get",
+        return_value=mock_device,
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.1.1.1",
+                "username": "test-username",
+                "password": "test-password",
+            },
+        )
+        await hass.async_block_till_done()
+
+    assert result2["type"] == FlowResultType.ABORT
+    assert result2["reason"] == "already_configured"
+
+
+async def test_form_invalid_auth(hass: HomeAssistant) -> None:
+    """Test we handle invalid auth."""
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": config_entries.SOURCE_USER}
+    )
+
+    with patch(
+        "pyecoforest.api.EcoforestApi.get",
+        side_effect=EcoforestAuthenticationRequired("401"),
+    ):
+        result2 = await hass.config_entries.flow.async_configure(
+            result["flow_id"],
+            {
+                "host": "1.1.1.1",
+                "username": "test-username",
+                "password": "test-password",
+            },
+        )
+
+    assert result2["type"] == FlowResultType.FORM
+    assert result2["errors"] == {"base": "invalid_auth"}
```

Comment:
Please end every config flow test with either a CONFIG_ENTRY or ABORT to show the config flow is able to recover from an exception

Reply:
My apologies I am not sure what do you mean with it, didn't seen it in any example :/
Can you help here?

Reply:
Not every integration has it, but we try to get it in every new integration now. The idea is that instead of only returning a form, you also finish the form. So if you return a form with errors, then call the function again with right data (and patch it correctly without side effect) and then test if it completes. That link for lastfm in the other thread shows an example:
https://github.com/home-assistant/core/blob/05254547380e95a902e41badc2c6159f250e11e1/tests/components/lastfm/test_config_flow.py#L58